### PR TITLE
Time-based rules: allow custom time via identifier

### PR DESCRIPTION
**Pull request by DFoxinator**
@be-dmitry 
@bmax 
@ingluisjimenez

### CODE REVIEW

File: src/Rules/TimeRuleAbstract.php [hunk 1]
```diff
@@ -13,4 +17,27 @@ protected function _getCurrentTime() {
...
+   * @param array $identifiers
+   *
+   * @return \DateTimeImmutable
+   *
+   * @throws \Behance\NBD\Gatekeeper\Exceptions\InvalidIdentifierException
+   */
+  protected function _getTimeFromIdentifiers( array $identifiers = [] ) {
+
+    if ( !isset( $identifiers[ self::IDENTIFIER_TYPE ] ) ) {
+      return $this->_getCurrentTime();
```
Comment (be-dmitry) [inline src/Rules/TimeRuleAbstract.php]: Looks liek this line is never getting his in your unit tests

**Comment by be-dmitry** [comment]
👍

File: src/Rules/TimeRuleAbstract.php (hunk 1 above)
Comment (DFoxinator) [inline src/Rules/TimeRuleAbstract.php]: this gets hit in the first set of tests that call canAccess with no identifiers

File: src/Rules/TimeRuleAbstract.php [hunk 2]
```diff
@@ -13,4 +17,33 @@ protected function _getCurrentTime() {
...
+   */
+  protected function _getTimeFromIdentifiers( array $identifiers = [] ) {
+
+    if ( !isset( $identifiers[ self::IDENTIFIER_TYPE ] ) ) {
+      return false;
+    }
+
+    $identifier = $identifiers[ self::IDENTIFIER_TYPE ];
+
+    if ( !is_object( $identifier ) || !( $identifier instanceof \DateTimeImmutable ) ) {
```
Comment (ingluisjimenez) [inline src/Rules/TimeRuleAbstract.php]: no need for checking if it's an object before instanceof

File: tests/unit/Rules/EndTimeIdentifierRuleTest.php [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+   *
+   * @param \DateTimeImmutable $end_time
+   * @param array              $identifiers
+   * @param bool               $expected
+   *
+   * @dataProvider timeIdentifierProvider
+   */
+  public function canAccessWithValidTimeIdentifier( \DateTimeImmutable $end_time, array $identifiers, $expected ) {
+
+    $rule = $this->getMock( EndTimeIdentifierRule::class, null, [ $end_time ] );
```
Comment (ingluisjimenez) [inline tests/unit/Rules/EndTimeIdentifierRuleTest.php]: need a mock for this? seems like no methods are being mocked. Same for other tests where a mock is being used

File: tests/unit/Rules/StartTimeRuleTest.php [hunk 4]
```diff
@@ -2,6 +2,7 @@
 
 namespace Behance\NBD\Gatekeeper\Rules;
 
+use Behance\NBD\Gatekeeper\Exceptions\InvalidIdentifierException;
```
Comment (ingluisjimenez) [inline tests/unit/Rules/StartTimeRuleTest.php]: using this namespace here?

File: src/Rules/TimeRuleAbstract.php [hunk 5]
```diff
@@ -13,4 +17,33 @@ protected function _getCurrentTime() {
...
+      throw new InvalidIdentifierException( '"time" identifier must be a DateTimeImmutable object.' );
+    }
+
+    return $identifier;
+
+  } // _getTimeFromIdentifiers
+
+  /**
+   * @param \DateTimeImmutable $time
+   * @return bool
```
Comment (ingluisjimenez) [inline src/Rules/TimeRuleAbstract.php]: nitpick: empty line before `@return`

File: src/Rules/TimeRuleAbstract.php [hunk 6]
```diff
@@ -13,4 +17,33 @@ protected function _getCurrentTime() {
...
+  /**
+   * @param array $identifiers
+   *
+   * @return \DateTimeImmutable|false
+   *
+   * @throws \Behance\NBD\Gatekeeper\Exceptions\InvalidIdentifierException
+   */
+  protected function _getTimeFromIdentifiers( array $identifiers = [] ) {
+
+    if ( !isset( $identifiers[ self::IDENTIFIER_TYPE ] ) ) {
```
Comment (ingluisjimenez) [inline src/Rules/TimeRuleAbstract.php]: the `identifiers` as array is starting to get more complex

**Comment by ingluisjimenez** [comment]
👍

File: tests/unit/Rules/EndTimeIdentifierRuleTest.php (hunk 3 above)
Comment (DFoxinator) [inline tests/unit/Rules/EndTimeIdentifierRuleTest.php]: good call

File: src/Rules/TimeRuleAbstract.php (hunk 6 above)
Comment (DFoxinator) [inline src/Rules/TimeRuleAbstract.php]: Yeah, maybe we can start to think about using an object here.

**Pull request merged**